crates/ruff_linter/src/rules/ruff/rules/invalid_rule_code.rs

```diff
@@ -0,0 +1,159 @@
+use crate::noqa::{Code, Directive};
+use crate::registry::Rule;
+use crate::Locator;
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_text_size::Ranged;
+
+use crate::noqa::{Codes, NoqaDirectiveLine, NoqaDirectives};
+
+/// ## What it does
+/// Checks for `noqa` codes that are invalid.
+///
+/// ## Why is this bad?
+/// Invalid rule codes serve no purpose and may indicate outdated code suppressions.
+///
+/// ## Example
+/// ```python
+/// import os  # noqa: XYZ999
+/// ```
+///
+/// Use instead:
+/// ```python
+/// import os
+/// ```
+///
+/// Or if there are still valid codes needed:
+/// ```python
+/// import os  # noqa: E402
+/// ```
+
+#[derive(ViolationMetadata)]
+pub(crate) struct InvalidRuleCode {
+    pub(crate) rule_code: String,
+}
+
+impl AlwaysFixableViolation for InvalidRuleCode {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Invalid rule code in `# noqa`: {}", self.rule_code)
+    }
+
+    fn fix_title(&self) -> String {
+        "Remove the rule code".to_string()
+    }
+}
+
+/// RUF102 for invalid noqa codes
+pub(crate) fn invalid_noqa_code(
+    diagnostics: &mut Vec<Diagnostic>,
+    noqa_directives: &NoqaDirectives,
+    locator: &Locator,
+    external: &[String],
+) {
+    for line in noqa_directives.lines() {
+        let Directive::Codes(directive) = &line.directive else {
+            continue;
+        };
+
+        let mut invalid_code_refs = vec![];
+        let mut all_invalid = true;
+
+        for code in directive.iter() {
+            let code_str = code.as_str();
+            if external.iter().any(|ext| code_str.starts_with(ext))
+                || Rule::from_code(code.as_str()).is_ok()
+            {
+                all_invalid = false;
+            } else {
+                invalid_code_refs.push(code);
+            }
+        }
+
+        if invalid_code_refs.is_empty() {
+            continue;
+        }
+
+        if all_invalid {
+            handle_all_codes_invalid(diagnostics, directive);
+        } else {
+            let valid_codes = directive
+                .iter()
+                .filter_map(|code| {
+                    let code_str = code.as_str();
+                    if external.iter().any(|ext| code_str.starts_with(ext))
+                        || Rule::from_code(code_str).is_ok()
+                    {
+                        Some(code_str)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+            handle_some_codes_invalid(diagnostics, &invalid_code_refs, &valid_codes, line, locator);
+        }
+    }
+}
+
+fn handle_all_codes_invalid(diagnostics: &mut Vec<Diagnostic>, directive: &Codes<'_>) {
```

Comment:
Nit: To reduce the number of arguments:
```suggestion
fn create_all_codes_invalid_diagnostic(directive: &Codes<'_>) -> Diagnostic {
```

Reply:
Adressed in [921f066](https://github.com/astral-sh/ruff/pull/17138/commits/921f0663f93257af10cdc8a5aec7fb7f34bad877)